File: 15.js
require('./dirty-tricks')
const {sum} = require("./util");
const input = require('fs').readFileSync('./examples/15.txt', 'utf-8')

const inputs = input.split(',')

function hash(string) {
  return string.reduce((acc, c) => ((acc + c.charCodeAt(0)) * 17) % 256, 0)
}

console.log('solution1', inputs.map(hash).reduce(sum))

const boxes = inputs
  .map(c => c.match(/(\w+)([=-])(\d?)/).slice(1, 4))
  .reduce((acc, [label, op, fl]) => {
    const h = hash(label)
    if (!acc[h]) acc[h] = {}
    if (op === '-') {
      delete acc[h][label]
    } else if (op === '=') {
      acc[h][label] = fl
    }
    return acc
  }, {})  

console.log(boxes)

const solution2 = Object.entries(boxes).reduce((acc, [box, contents]) => {
  const boxNum = parseInt(box)
  return acc + Object.entries(contents).reduce((acc, [label, fl], i) => {
    return acc + (boxNum + 1) * (i + 1) * parseInt(fl)
  }, 0)
}, 0)

console.log('solution2', solution2)

File: 15.exs
input =
  File.read!("./examples/15.txt")
  |> String.split(",")

hash = fn string ->
  String.to_charlist(string)
  |> Enum.reduce(0, fn val, acc -> rem((acc + val) * 17, 256) end)
end

input
|> Enum.map(hash)
|> Enum.sum()
|> then(&IO.puts("solution1 = #{&1}"))

input
|> Enum.map(&Regex.run(~r/(\w+)([=-])(\d?)/, &1))
|> Enum.reduce(%{}, fn [_, label, op, fl], acc ->
  h = hash.(label)
  v = Map.get(acc, h, [])

  Map.put(
    acc,
    h,
    case op do
      "=" ->
        if Enum.any?(v, fn {l, _} -> l == label end) do
          Enum.map(v, fn {l, ofl} ->
            if l == label, do: {l, fl}, else: {l, ofl}
          end)
        else
          v ++ [{label, fl}]
        end

      "-" ->
        Enum.filter(v, fn {l, _} -> l != label end)
    end
  )
end)
|> Enum.reduce(0, fn {h, list}, acc ->
  Enum.with_index(list)
  |> Enum.reduce(0, fn {{_, fl}, index}, acc ->
    acc + (h + 1) * (index + 1) * String.to_integer(fl)
  end)
  |> then(&(acc + &1))
end)
|> then(&IO.puts("solution2 = #{&1}"))
